log4j-core-test/src/test/java/org/apache/logging/log4j/core/pattern/ThrowableFilterTest.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.pattern;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.List;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.test.appender.ListAppender;
+import org.apache.logging.log4j.core.test.junit.LoggerContextSource;
+import org.apache.logging.log4j.core.test.junit.Named;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Unit tests for {@code throwable} pattern.
+ */
+@LoggerContextSource("log4j-throwable-filter.xml")
+public class ThrowableFilterTest {
```

Comment:
>     2. It programmatically creates a `Configuration` and a `LoggerContext`:

I think this is a step in the right direction, since it replaces a full-blown integration test with an almost unit test, but this approach is still computationally expensive.

The `Configuration` argument of `ThrowablePatternConverter.newInstance()` and other `newInstance()` methods is **optional**. We can simply use:

```java
final Throwable thrown = new RuntimeException();
final ThrowableProxy thrownProxy = new ThrowableProxy(thrown);
final LogEvent event = Log4jLogEvent.newBuilder();
    .setThrown(thrown)
    .setThrownProxy(thrownProxy)
    .build();
final StringBuilder toAppendTo = new StringBuilder();
final LogEventPatternConverter converter = ThrowablePatternConverter.newInstance(null, options);
converter.format(event, toAppendTo);
// Some assertions
```

Reply:
@ppkarwasz, we are not computationally worsening the situation. Each test is getting a fresh `LoggerContext`, as each `@SpringBootTest` would get a new `ApplicationContext`.

Plus, I don't think your suggestion works for properties passed at the root `Configuration` level – unless we create an explicit `Configuration`.

IIRC, we agreed that all such static `Component.newInstance()` methods should be gone in Log4j 3, since they all create an implicit `DefaultConfiguration`. There is only one way to create a bean: request one from the dependency injection framework. The other alternative is to call its ctor with all its args mocked.

In short, I am in favor of sticking to the `ConfigurationBuilder`, since it has all the benefits and not worsening the computation complexity.

Reply:
My point is that we don't need neither a `LoggerContext` nor a `Configuration` to test the throwable converters.
As far as I can tell, the only `Configuration` method that is used is the deprecated `getPluginPackages()`.
If the `Configuration` parameter is missing `DefaultConfiguration` is **not** used:

https://github.com/apache/logging-log4j2/blob/9629d51931dffdd85b4e9bbe326f42b6d6565c08/log4j-core/src/main/java/org/apache/logging/log4j/core/layout/PatternLayout.java#L260-L271

https://github.com/apache/logging-log4j2/blob/9629d51931dffdd85b4e9bbe326f42b6d6565c08/log4j-core/src/main/java/org/apache/logging/log4j/core/pattern/PatternParser.java#L127-L134

Reply:
I agree creating test programmatically is pretty cool!
And both approach seems good to me. Using `Configuration` approach we can build the converter in more like top-level `xml/yaml` way. Using `newInstance` is more like low-level API approach.
I would prefer `Configuration` since it's more straight forward because I can create tests following the docs.
But this is just my thought, I am not familiar with whole codebase 💦

Reply:
@alan0428a, please proceed with the `ConfigurationBuilder` approach. I will internally talk with @ppkarwasz and if changes are needed, we will carry them out ourselves later.

Reply:
Just found there are actually only 6 combinitions, 3 types of converter x (package, depth). `filters == package`
But maybe we can add default case (no special setting of`packagel` &  `depth`) for each convert test case?

Reply:
@vy Not sure why the programaically created `ListAppender.raw` is always true, so it writes the event to `data` instead of `message`. The `ListAppender` created through xml has the `raw` as `false`. Could you guide me on this?

Reply:
You should probably remove the `addAttribute("raw", true)` in Volkan's code, so you can compare strings, not `byte[]`s.

Reply:
@ppkarwasz 
Ahh...make sense, didn't notice that attribute 😅. Thanks